{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "T5cHxzDGk7k5",
        "outputId": "20a60420-a1d0-4bcf-dfd3-81daef1e4e98"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: CCTop in /usr/local/lib/python3.11/site-packages (1.0.1)\n",
            "âœ¨ðŸ°âœ¨ Everything looks OK!\n",
            "Channels:\n",
            " - bioconda\n",
            " - conda-forge\n",
            "Platform: linux-64\n",
            "Collecting package metadata (repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\bdone\n",
            "Solving environment: / \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "    current version: 24.11.3\n",
            "    latest version: 25.5.1\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c conda-forge conda\n",
            "\n",
            "\n",
            "\n",
            "# All requested packages already installed.\n",
            "\n",
            "Channels:\n",
            " - bioconda\n",
            " - conda-forge\n",
            "Platform: linux-64\n",
            "Collecting package metadata (repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "Solving environment: \\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "    current version: 24.11.3\n",
            "    latest version: 25.5.1\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c conda-forge conda\n",
            "\n",
            "\n",
            "\n",
            "# All requested packages already installed.\n",
            "\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.11/site-packages (2.3.1)\n",
            "Requirement already satisfied: numpy>=1.23.2 in /usr/local/lib/python3.11/site-packages (from pandas) (2.3.1)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.11/site-packages (from pandas) (2.9.0.post0)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/site-packages (from pandas) (2025.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/site-packages (from pandas) (2025.2)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/site-packages (from python-dateutil>=2.8.2->pandas) (1.17.0)\n",
            "bed file: /content/batch_sgRNA/test/ncrna_teste.bed\n",
            "extended size on the chromosome position: 100\n",
            "extended size on the target position: 50\n",
            "Reference FASTA: /content/batch_sgRNA/test/LBRAZ_M2903.Dec2022.fasta\n",
            "output bed: targetprimer\n",
            "Extract fasta of initial chomossomal position of ncRNA\n",
            "Extract fasta of end position of ncRNA\n",
            "Extraction of fasta from new position according to target\n",
            "skipping samtools\n",
            "Running cctop\n",
            "Working on sequence 'LbrM2903_01_v2_pilon_lncRNA21LbrM2903_01_v2_pilon104733-104883'\n",
            "Working on sequence 'LbrM2903_01_v2_pilon_lncRNA22LbrM2903_01_v2_pilon108628-108778'\n",
            "Working on sequence 'LbrM2903_01_v2_pilon_lncRNA30LbrM2903_01_v2_pilon128449-128599'\n",
            "Working on sequence 'LbrM2903_05_v2_pilon_lncRNA1LbrM2903_05_v2_pilon4796-4946'\n",
            "Working on sequence 'LbrM2903_05_v2_pilon_lncRNA2LbrM2903_05_v2_pilon5601-5751'\n",
            "Working on sequence 'LbrM2903_05_v2_pilon_lncRNA3LbrM2903_05_v2_pilon8372-8522'\n",
            "Working on sequence 'LbrM2903_29_v2_pilon_lncRNA114LbrM2903_29_v2_pilon622503-622653'\n",
            "Working on sequence 'LbrM2903_12_v2_pilon_lncRNA46LbrM2903_12_v2_pilon234770-234920'\n",
            "Working on sequence 'LbrM2903_30_v2_pilon_lncRNA114LbrM2903_30_v2_pilon555628-555778'\n",
            "Working on sequence 'LbrM2903_22_v2_pilon_lncRNA116LbrM2903_22_v2_pilon632157-632307'\n",
            "Working on sequence 'LbrM2903_35_v2_pilon_lncRNA228LbrM2903_35_v2_pilon1598785-1598935'\n",
            "Working on sequence 'LbrM2903_01_v2_pilon_lncRNA21LbrM2903_01_v2_pilon104447-104597'\n",
            "Working on sequence 'LbrM2903_01_v2_pilon_lncRNA22LbrM2903_01_v2_pilon107288-107438'\n",
            "Working on sequence 'LbrM2903_01_v2_pilon_lncRNA30LbrM2903_01_v2_pilon128195-128345'\n",
            "Working on sequence 'LbrM2903_05_v2_pilon_lncRNA1LbrM2903_05_v2_pilon4283-4433'\n",
            "Working on sequence 'LbrM2903_05_v2_pilon_lncRNA2LbrM2903_05_v2_pilon4750-4900'\n",
            "Working on sequence 'LbrM2903_05_v2_pilon_lncRNA3LbrM2903_05_v2_pilon6973-7123'\n",
            "Working on sequence 'LbrM2903_29_v2_pilon_lncRNA114LbrM2903_29_v2_pilon621928-622078'\n",
            "Working on sequence 'LbrM2903_12_v2_pilon_lncRNA46LbrM2903_12_v2_pilon234397-234547'\n",
            "Working on sequence 'LbrM2903_30_v2_pilon_lncRNA114LbrM2903_30_v2_pilon555264-555414'\n",
            "Working on sequence 'LbrM2903_22_v2_pilon_lncRNA116LbrM2903_22_v2_pilon631798-631948'\n",
            "Working on sequence 'LbrM2903_35_v2_pilon_lncRNA228LbrM2903_35_v2_pilon1598340-1598490'\n",
            "pos-processing output cctop for all genes\n",
            "We have the best sgRNAs: cctop_listtop3.tsv\n"
          ]
        }
      ],
      "source": [
        "import os\n",
        "!git clone https://github.com/AKCLAB/batch_sgRNA.git\n",
        "os.chdir(\"/content/batch_sgRNA/\") # Set this directory\n",
        "#Another way is to create a shortcut to the batch_sgRNA directory in your own Google Drive and access it this way.\n",
        "#from google.colab import drive\n",
        "#drive.mount('/content/drive')\n",
        "#!ls \"/content/drive/MyDrive/batch_sgRNA\"\n",
        "!pip install CCTop\n",
        "!pip install -q condacolab\n",
        "import condacolab\n",
        "condacolab.install()\n",
        "!conda install -y -c bioconda bowtie\n",
        "!conda install -y -c bioconda bedtools\n",
        "!pip install pandas\n",
        "\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "#!which bowtie-build\n",
        "#!which cctop\n",
        "# The test directory contains three reference genomes of *Leishmania braziliensis* (MHOM/BR/75/M2903):\n",
        "# - Beverley version (LBRAZ_M2903.Dec2022.fasta)\n",
        "# - TriTrypDB version 30 (TriTrypDB-30_LbraziliensisMHOMBR75M2903_Genome.fasta)\n",
        "# - TriTrypDB version 68 (TriTrypDB-68_LbraziliensisMHOMBR75M2903_Genome.fasta)\n",
        "\n",
        "# It also includes the file 'ncrna_teste.bed', which contains the start and end coordinates of the target genes.\n",
        "# df = pd.read_csv(\"/content/batch_sgRNA/test/ncrna_teste.bed\", sep = \"\\t\") # Confer files\n",
        "\n",
        "!bash batch_sgrna.sh -bedfile /content/batch_sgRNA/test/ncrna_teste.bed -reffasta /content/batch_sgRNA/test/LBRAZ_M2903.Dec2022.fasta -size_target 50 -size_chr 100"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
